FWCore/Framework/src/ProductResolvers.cc

```diff
@@ -430,14 +430,6 @@ namespace edm {
     }
   }
 
-  void PuttableProductResolver::putProduct(std::unique_ptr<WrapperBase> edp) const {
-    ProducedProductResolver::putProduct(std::move(edp));
-    bool expected = false;
-    if (prefetchRequested_.compare_exchange_strong(expected, true)) {
-      m_waitingTasks.doneWaiting(std::exception_ptr());
```

Comment:
The other `m_waitingTasks.doneWaiting` call requires that `worker_` not be `nullptr`. Is there any condition under which we do have it as `nullptr`?

Reply:
The `worker_` is initially uninitialized (already which sort of defeats the purpose of the check...) and is set later by
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/FWCore/Framework/src/ProductResolvers.cc#L449-L451

The `setupUnscheduled()` is called by `Principal::setupUnscheduled()`
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/FWCore/Framework/src/Principal.cc#L398-L400
that is called by
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/FWCore/Framework/src/WorkerManager.cc#L141-L148

That in turn is called from `GlobalSchedule::processOneGlobalAsync()`
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/FWCore/Framework/interface/GlobalSchedule.h#L220-L222
, from `StreamSchedule::processOneEventAsync()`
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/FWCore/Framework/src/StreamSchedule.cc#L915
and from
https://github.com/cms-sw/cmssw/blob/3cac6aaba83c4b00b452f8fee244e5bccafc809b/Mixing/Base/src/SecondaryEventProvider.cc#L121-L127
(the `SecondaryEventProvider::setupPileUpEvent()` ends up being called from `PreMixingModule` and `DataMixingModule` once for each pileup event)

To me it seems the workers are being set for the ProductResolvers before anything gets run (although I wonder why the setup gets repeated), and I can't see why anything would call `putProduct()` without `worker_` being set (but I could be missing some corner case).